google-cloud-spanner/src/main/java/com/google/cloud/spanner/AsyncResultSet.java

```diff
@@ -0,0 +1,226 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.spanner;
+
+import com.google.api.core.ApiFuture;
+import com.google.common.base.Function;
+import com.google.common.collect.ImmutableList;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executor;
+
+/** Interface for result sets returned by async query methods. */
+public interface AsyncResultSet extends ResultSet {
+
+  /**
+   * Interface for receiving asynchronous callbacks when new data is ready. See {@link
+   * AsyncResultSet#setCallback(Executor, ReadyCallback)}.
+   */
+  public static interface ReadyCallback {
+    CallbackResponse cursorReady(AsyncResultSet resultSet);
+  }
+
+  /** Response code from {@code tryNext()}. */
+  public enum CursorState {
+    /** Cursor has been moved to a new row. */
+    OK,
+    /** Read is complete, all rows have been consumed, and there are no more. */
+    DONE,
+    /** No further information known at this time, thus current row not available. */
+    NOT_READY
+  }
+
+  /**
+   * Non-blocking call that attempts to step the cursor to the next position in the stream. The
+   * cursor may be inspected only if the cursor returns {@code CursorState.OK}.
+   *
+   * <p>A caller will typically call {@link #tryNext()} in a loop inside the ReadyCallback,
+   * consuming all results available. For more information see {@link #setCallback(Executor,
+   * ReadyCallback)}.
+   *
+   * <p>Currently this method may only be called if a ReadyCallback has been registered. This is for
+   * safety purposes only, and may be relaxed in future.
+   *
+   * @return current cursor readiness state
+   * @throws SpannerException When an unrecoverable problem downstream occurs. Once this occurs you
+   *     will get no further callbacks. You should return CallbackResponse.DONE back from callback.
+   */
+  CursorState tryNext() throws SpannerException;
+
+  /**
+   * Register a callback with the ResultSet to be made aware when more data is available, changing
+   * the usage pattern from sync to async. Details:
+   *
+   * <ul>
+   *   <li>The callback will be called at least once.
+   *   <li>The callback is run each time more results are available, or when we discover that there
+   *       will be no more results. (unless paused, see below). Spurious callbacks are possible, see
+   *       below.
+   *   <li>Spanner guarantees that one callback is ever outstanding at a time. Also, future
+   *       callbacks guarantee the "happens before" property with previous callbacks.
+   *   <li>A callback normally consumes all available data in the ResultSet, and then returns {@link
+   *       CallbackResponse#CONTINUE}.
+   *   <li>If a callback returns {@link CallbackResponse#CONTINUE} with data still in the ResultSet,
```

Comment:
The documentation is (largely) copy-paste from the design document, but there is a (very small) performance penalty for just returning `CONTINUE` instead of looping over all available results. The `AsyncResultSetImpl` implementation will check whether the result set has been cancelled when the callback returns with `CONTINUE`. To do so, it needs to take an internal lock.

Reply:
I see, good to know, thanks!